src/Streams/FileReader.cpp

```diff
@@ -56,7 +57,17 @@ namespace Stream
 		file.seekg(position);
 	}
 
-	void FileReader::SeekRelative(int64_t offset) {
+	void FileReader::SeekRelative(int64_t offset) 
+	{
+		if (offset < 0) {
+			const int mask = offset >> ( sizeof(int64_t) * CHAR_BIT - 1 );
+			const uint64_t offsetAbsValue = (offset ^ mask) - mask;
```

Comment:
Ahh, `-x` is the same as flip all the bits and add 1, which is the same as XOR with -1 and then add -1. The first part is sign bit replication to the full value. A negative number replicates the sign bit (1) to all bits, which results in -1. A position number replicates the sign bit (0) to all bits, which results in 0. And XOR with 0 does not, and a subtraction of 0 does nothing. Hence it performs `-x` for negative `x`, and is a no-op for positive values. All branch free. Clever.

Though certainly not very clear. I'd be hesitant to use that without wrapping it in a function. It's at a much lower abstraction than the surrounding code.

Mind you, there is already a [std::abs](https://en.cppreference.com/w/cpp/numeric/math/abs) function in the standard library, with overloads for various integer types.

Still, that was fun to work through.

Reply:
Good point, well now I feel dumb. Sorry about that.

Besides the odd implementation, do you think this check is a good idea? I think currently the MSVC implementation of ifstream's position will just wrap to a really large positive number if you seek before zero. I think seeking before the start of a file would constitute an error and should throw.

I might also be willing to just ditch allowing a signed relative offset here in favor of an unsigned value. I thought it would be nice to allow others to seek backwards, but it seems to cause a lot of trouble. Removing the ability to use a signed value would remove the check altogether.

-Brett

Reply:
We should bounds check things.

I don't think using an `unsigned` value really avoids the problem. Both signed and unsigned values have corresponding equivalent bit representations, which will produce the same result after the addition. Or at least that's true where signed values have well defined wrap around, which the C++ standard doesn't guarantee, but it nevertheless common.

That is, a `SeekRelative(-1)` should produce the same offset as a `SeekRelative(std::numeric_limits<uint64_t>::max - 1)`. At least that's true in the absence of error checks to prevent it. I suppose with the assumption of positive values, that would be considered an invalid offset, and rejected as too large.

I kind of wanted to stick with `unsigned` values for exactly this reason, and also because they have well defined wrap around. Plus, to seek to the end of a really long stream, you might actually need to full range of positive values. In practice we're unlikely to encounter a stream that long though. And really, a `SeekRelative` is probably expected to allow for negative values. I think it's more correct from an API standpoint to allow for negative offset seeks.

Unless of course you want a separate method to allow seeking backward? Hmm, that might actually make sense. It's probably unlikely you'd want a variable seek distance that could go either forward or backward. You'd probably choose ahead of time to either go forward or backward, or just do an absolute seek. Maybe we should make it part of the method name. That also solves the range issue when you need to do a seek all the way across a long stream. I think this is similar to why I wanted separate methods for `Seek` and `SeekRelative`, rather than a parameter to control the initial seek point (which inevitably ends up being a hardcoded constant most of the time).